L1Trigger/TrackFindingTracklet/interface/DRin.h

```diff
@@ -0,0 +1,155 @@
+#ifndef L1Trigger_TrackFindingTracklet_DRin_h
+#define L1Trigger_TrackFindingTracklet_DRin_h
+
+#include "L1Trigger/TrackTrigger/interface/Setup.h"
+#include "L1Trigger/TrackerTFP/interface/DataFormats.h"
+#include "L1Trigger/TrackerTFP/interface/LayerEncoding.h"
+#include "L1Trigger/TrackFindingTracklet/interface/ChannelAssignment.h"
+#include "L1Trigger/TrackFindingTracklet/interface/Settings.h"
+
+#include <vector>
+#include <deque>
+
+namespace trklet {
+
+  /*! \class  trklet::DRin
+   *  \brief  Class to emulate transformation of tracklet tracks and stubs into TMTT format
+   *          and routing of seed type streams into inv2R streams
+   *  \author Thomas Schuh
+   *  \date   2023, Jan
+   */
+  class DRin {
+  public:
+    DRin(const edm::ParameterSet& iConfig,
+         const tt::Setup* setup_,
+         const trackerTFP::DataFormats* dataFormats,
+         const trackerTFP::LayerEncoding* layerEncoding,
+         const ChannelAssignment* channelAssignment,
+         const Settings* settings,
+         int region);
+    ~DRin() {}
+    // read in and organize input tracks and stubs
+    void consume(const tt::StreamsTrack& streamsTrack, const tt::StreamsStub& streamsStub);
+    // fill output products
+    void produce(tt::StreamsStub& accpetedStubs,
+                 tt::StreamsTrack& acceptedTracks,
+                 tt::StreamsStub& lostStubs,
+                 tt::StreamsTrack& lostTracks);
+
+  private:
+    // truncates double precision of val into base precision, +1.e-12 restores robustness of addition of 2 digitised values
+    double digi(double val, double base) const { return (floor(val / base + 1.e-12) + .5) * base; }
+    // basetransformation of val from baseLow into baseHigh using widthMultiplier bit multiplication
+    double redigi(double val, double baseLow, double baseHigh, int widthMultiplier) const;
+    struct Stub {
+      Stub(const TTStubRef& ttStubRef, int layer, int layerDet, int stubId, double r, double phi, double z, bool psTilt)
+          : valid_(true),
+            ttStubRef_(ttStubRef),
+            layer_(layer),
+            layerDet_(layerDet),
+            stubId_(stubId),
+            r_(r),
+            phi_(phi),
+            z_(z),
+            psTilt_(psTilt) {}
+      bool valid_;
+      TTStubRef ttStubRef_;
+      //
+      int layer_;
+      // layer id [0-5] barrel [6-10] end cap discs
```

Comment:
Is this using the KF format and not layers in the format of 4 barrel and 4 disk stubs (excluding seeds)?

Reply:
layer_ is at construction time this id: [barrel: 1-6, discs: 11-15] using seed type and channel number defined here [https://github.com/cms-L1TK/cmssw/blob/tschuh_dr/L1Trigger/TrackFindingTracklet/python/ChannelAssignment_cfi.py#L27-L36](https://github.com/cms-L1TK/cmssw/blob/tschuh_dr/L1Trigger/TrackFindingTracklet/python/ChannelAssignment_cfi.py#L27-L36). Later it gets overwritten with kf layer id [0-6].

Reply:
I just added a further member called layerKF_ to avoid confusion